_posts/2019-08-18-Securing-Web-Applications-With-Keycloak.md

```diff
@@ -0,0 +1,608 @@
+
+```typescript
+export class HeroService {
+
+  private heroesUrl = 'http://localhost:9090/api/heroes';
+  ...
+}
+```
+
+It will still try to fetch preloaded heroes from memory, so delete the `HttpClientInMemoryWebApiModule` from `app.module.ts`.
+
+Now you can run the application with
+
+```bash
+$ ng serve
+```
+
+and browse to [localhost:4200](http://localhost:4200) to see if it works.
+
+> Please note that not all functionality of the app is working because - for brevity - we only implemented the GET functionality in the Spring Boot back end.
+
+# Implementing Security
+We will now dive into the interesting part: setting up the security of the applications.
+To understand this section, you should have a basic understanding of OAuth 2.0 and OpenID Connect (OIDC). 
+If this is not yet the case, this section might be more difficult to understand. 
+You can fast-forward to the next section or start your journey by watching this awesome video by [Nate Barbettini](https://twitter.com/nbarbettini) from [Okta](https://www.okta.com/).
+
+<div class="responsive-video">
+    <iframe width="1164" height="655" src="https://www.youtube.com/embed/996OiexHze0" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>
+</div>
+<br>
+
+## Implicit Flow versus Code Flow + PKCE
+In this example, we will use the authorization code grant flow with Proof Key for Code Exchange ([PKCE](https://tools.ietf.org/html/rfc7636)) to secure the Angular app. 
+It's a very long name for what could be shortened to "code flow + PKCE" which is [more secure than the implicit flow](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-13#section-3.1.2).   
+In fact, the implicit flow was never very secure to begin with. 
+This is well-explained by [Nate Barbettini](https://twitter.com/nbarbettini) and [Aaron Parecki](https://twitter.com/aaronpk
+):
+
+<div class="responsive-video">
+    <iframe width="1164" height="655" src="https://www.youtube.com/embed/CHzERullHe8" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>
+</div>
+<br>
+
+The implicit flow was the easiest to understand, since it required one step less than the standard code flow:
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/implicit_vs_code.png" width="60%">
+</div>
+
+PKCE is an addition on top of the standard code flow to make it usable for public clients. 
+It is already in use for native and mobile clients. 
+PKCE boils down to this: 
+1. Give `hash` of random value to authorization server when logging in to ask for `code`
+2. Hand over the `random value` to authorization server when exchanging `code` for `access token`
+3. Authorization server returns `access token` after verifying that `hash` belongs to `random value`.
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/pkce.png" width="70%">
```

Comment:
Add an image with eg. `<img class="image fit" src="{{ '/img/devoxx-ma-2018/speakers-dinner.jpg' | prepend: site.baseurl }}" alt="Speakers Dinner" width="60%" />`, otherwise this won't scale good.